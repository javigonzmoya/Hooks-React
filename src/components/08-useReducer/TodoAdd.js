import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
  const [values, handleInputChange, reset] = useForm({ desciption: '' });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (values.desciption.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: values.desciption,
      done: false,
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="desciption"
          className="form-control"
          placeholder="Hacer.."
          autoComplete="off"
          onChange={handleInputChange}
          value={values.desciption}
        />
        <div className="d-grid gap-2">
          <button className="btn btn-outline-primary mt-1 " type="submit">
            Agregar
          </button>
        </div>
      </form>
    </>
  );
};
