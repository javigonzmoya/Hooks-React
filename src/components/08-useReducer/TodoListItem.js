import React from 'react';

export const TodoListItem = ({ todo, index, handleRemove, handleToggle }) => {
  const { id, desc, done } = todo;
  return (
    <li className="list-group-item">
      <p
        className={`${done ? 'complete' : ''} text-center`}
        onClick={() => handleToggle(id)}
      >
        {index + 1}.{desc}
      </p>
      <button className="btn btn-danger" onClick={() => handleRemove(id)}>
        borrar
      </button>
    </li>
  );
};
