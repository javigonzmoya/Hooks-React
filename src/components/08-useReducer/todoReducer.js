export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case 'add':
      return [...state, action.payload];
    case 'remove':
      return state.filter((s) => s.id !== action.payload);
    case 'toggle':
      return state.map((s) =>
        s.id === action.payload ? { ...s, done: !s.done } : s
      );
    case 'toggle-old':
      return state.map((s) => {
        if (s.id === action.payload) {
          return {
            ...s,
            done: !s.done,
          };
        }
        return s;
      });
    default:
      return state;
  }
};
