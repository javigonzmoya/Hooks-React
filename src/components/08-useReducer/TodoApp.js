import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import { todoReducer } from './todoReducer';

import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);
  const [values, handleInputChange, reset] = useForm({ desciption: '' });

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleSubmit = (e) => {
    e.preventDefault();

    if (values.desciption.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: values.desciption,
      done: false,
    };

    const action = {
      type: 'add',
      payload: newTodo,
    };
    dispatch(action);
    reset();
  };

  const handleRemove = (todoId) => {
    const action = {
      type: 'remove',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };
    dispatch(action);
  };
  return (
    <div>
      <h1>TodoApp: ( {todos.length} )</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            {todos.map(({ id, desc, done }, i) => (
              <li key={id} className="list-group-item">
                <p
                  className={`${done ? 'complete' : ''} text-center`}
                  onClick={() => handleToggle(id)}
                >
                  {i + 1}.{desc}
                </p>
                <button
                  className="btn btn-danger"
                  onClick={() => handleRemove(id)}
                >
                  borrar
                </button>
              </li>
            ))}
          </ul>
        </div>

        <div className="col-5">
          <h4>Agregar TODO</h4>
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              name="desciption"
              className="form-control"
              placeholder="Hacer.."
              autoComplete="off"
              onChange={handleInputChange}
              value={values.desciption}
            />
            <div className="d-grid gap-2">
              <button className="btn btn-outline-primary mt-1 " type="submit">
                Agregar
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
