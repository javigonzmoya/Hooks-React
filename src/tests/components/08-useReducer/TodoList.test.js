import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas sobre <TodoList/>', () => {
  const handleRemove = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(
    <TodoList
      todos={demoTodos}
      handleRemove={handleRemove}
      handleToggle={handleToggle}
    />
  );

  test('debe de mostrarse correctamente ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de tener dos <TodoListItem/> ', () => {
    const elemntos = wrapper.find('TodoListItem');
    expect(elemntos.length).toBe(demoTodos.length);
    expect(elemntos.at(0).prop('handleToggle')).toEqual(expect.any(Function));
    expect(elemntos.at(0).prop('handleRemove')).toEqual(expect.any(Function));
  });
});
