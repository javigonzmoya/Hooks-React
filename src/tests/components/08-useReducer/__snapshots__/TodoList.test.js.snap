// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <TodoList/> debe de mostrarse correctamente  1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleRemove={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender react",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleRemove={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Angular",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
